### PR TITLE
avahi: Support Conan V2

### PR DESCRIPTION
**Pull request by jwillikers**
Specify library name and version:  **avahi/***

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/avahi/all/conanfile.py [hunk 1]
```diff
@@ -26,75 +30,68 @@ class AvahiConan(ConanFile):
...
     def validate(self):
-        if self.settings.os != "Linux" or tools.cross_building(self):
+        if self.info.settings.os != "Linux":
             raise ConanInvalidConfiguration("Only Linux is supported for this package.")
 
     def configure(self):
-        del self.settings.compiler.cppstd
-        del self.settings.compiler.libcxx
+        try:
+            del self.settings.compiler.cppstd
```
Comment (prince-chrismc) [inline recipes/avahi/all/conanfile.py:50]: This is way better then doing them one at a time....

File: recipes/avahi/all/conanfile.py [hunk 2]
```diff
@@ -26,75 +30,68 @@ class AvahiConan(ConanFile):
...
+
+    def generate(self):
+        tc = AutotoolsToolchain(self)
+        tc.configure_args.append("--enable-compat-libdns_sd")
+        tc.configure_args.append("--disable-gtk3")
+        tc.configure_args.append("--disable-mono")
+        tc.configure_args.append("--disable-monodoc")
+        tc.configure_args.append("--disable-python")
+        tc.configure_args.append("--disable-qt5")
+        tc.configure_args.append("--with-systemdsystemunitdir=/lib/systemd/system")
```
Comment (prince-chrismc) [inline recipes/avahi/all/conanfile.py:71]: Noob question, why is it not relative to package folder? Should this be an option?

File: recipes/avahi/all/conanfile.py [hunk 3]
```diff
@@ -26,75 +30,68 @@ class AvahiConan(ConanFile):
...
+        rm(self, "*.la", os.path.join(self.package_folder, "lib"))
 
     def package_info(self):
         self.cpp_info.names["cmake_find_package"] = "Avahi"
         self.cpp_info.names["cmake_find_package_multi"] = "Avahi"
 
         for lib in ("client", "common", "core", "glib", "gobject", "libevent", "compat-libdns_sd"):
-            avahi_lib = "avahi-{}".format(lib)
+            avahi_lib = f"avahi-{lib}"
             self.cpp_info.components[lib].names["cmake_find_package"] = lib
```
Comment (prince-chrismc) [inline recipes/avahi/all/conanfile.py:95]: New names are missing

File: recipes/avahi/all/test_package/CMakeLists.txt [hunk 4]
```diff
@@ -1,10 +1,7 @@
...
-
-include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
-conan_basic_setup(TARGETS)
+project(test_package LANGUAGES C)
 
 find_package(Avahi CONFIG REQUIRED)
  
 add_executable(${PROJECT_NAME} test_package.c)
-target_link_libraries(${PROJECT_NAME} Avahi::compat-libdns_sd)
+target_link_libraries(${PROJECT_NAME} PRIVATE avahi::compat-libdns_sd)
```
Comment (prince-chrismc) [inline recipes/avahi/all/test_package/CMakeLists.txt:7]: Was it previously wrong?

**Review by prince-chrismc — APPROVED** [review]

**Review by prince-chrismc — COMMENTED** [review]
Rip I had questions + missing new properties for names

File: recipes/avahi/all/conanfile.py (hunk 1 above)
Comment (jwillikers) [reply recipes/avahi/all/conanfile.py:50]: ikr? :)

File: recipes/avahi/all/conanfile.py (hunk 2 above)
Comment (jwillikers) [reply recipes/avahi/all/conanfile.py:71]: It's how the package uses `DESTDIR` when installing. It ends up using appending this path to `DESTDIR` which is the package directory already. By incorporating the `self.package_folder`, the package directory path is duplicated and this ends up in a very strange place in the package folder. I believe the way the new `AutotoolsToolchain` handles `DESTDIR` is why this change was necessary. As for making it an option, I'm not sure that's necessary. Putting it the standard directory path seems to make the most sense right now since it keeps the previous behavior of the package. Maybe is someone specifically requests this be an option in the future, it can be added then.

File: recipes/avahi/all/test_package/CMakeLists.txt (hunk 4 above)
Comment (jwillikers) [reply recipes/avahi/all/test_package/CMakeLists.txt:7]: Yes, I believe it was by changing the default from `Avahi` to `avahi` for the old CMake generators. Avahi does not provide any sort of CMake Config nor does CMake provide an official Find Module. In this case, the defaults for the `CMakeDeps` generator should be used, right?

File: recipes/avahi/all/conanfile.py (hunk 3 above)
Comment (jwillikers) [reply recipes/avahi/all/conanfile.py:95]: Yes, this is intentional. Avahi does not provide CMake Config files, so we should actually be using the default names provided by the CMake Conan generators as far as I understand.

File: recipes/avahi/all/test_package/CMakeLists.txt (hunk 4 above)
Comment (uilianries) [reply recipes/avahi/all/test_package/CMakeLists.txt:7]: @jwillikers Correct!

File: recipes/avahi/all/conanfile.py [hunk 5]
```diff
@@ -15,7 +19,7 @@ class AvahiConan(ConanFile):
     homepage = "https://github.com/lathiat/avahi"
     license = "LGPL-2.1-only"
     settings = "os", "arch", "compiler", "build_type"
-    generators = "pkg_config"
+    generators = "AutotoolsDeps", "PkgConfigDeps"
```
Comment (uilianries) [inline recipes/avahi/all/conanfile.py:22]:
```suggestion
```
Avoid mixing generators attribute + method, for easier review.

File: recipes/avahi/all/conanfile.py [hunk 6]
```diff
@@ -1,8 +1,12 @@
-from conans import ConanFile, tools, AutoToolsBuildEnvironment
-from conans.errors import ConanInvalidConfiguration
 import os
 
-required_conan_version = ">=1.33.0"
+from conan import ConanFile
+from conan.tools.files import copy, get, rmdir, rm
+from conan.tools.gnu import Autotools, AutotoolsToolchain
```
Comment (uilianries) [inline recipes/avahi/all/conanfile.py:5]:
```suggestion
from conan.tools.gnu import Autotools, AutotoolsDeps, PkgConfigDeps, AutotoolsToolchain
```

File: recipes/avahi/all/conanfile.py [hunk 7]
```diff
@@ -1,8 +1,12 @@
...
+from conan.tools.layout import basic_layout
+from conan.errors import ConanInvalidConfiguration
+
+required_conan_version = ">=1.52.0"
```
Comment (uilianries) [inline recipes/avahi/all/conanfile.py:9]:
```suggestion
required_conan_version = ">=1.47.0"
```
Do not need to enforce to the latest version

File: recipes/avahi/all/conanfile.py [hunk 8]
```diff
@@ -26,75 +30,68 @@ class AvahiConan(ConanFile):
...
+    def generate(self):
+        tc = AutotoolsToolchain(self)
+        tc.configure_args.append("--enable-compat-libdns_sd")
+        tc.configure_args.append("--disable-gtk3")
+        tc.configure_args.append("--disable-mono")
+        tc.configure_args.append("--disable-monodoc")
+        tc.configure_args.append("--disable-python")
+        tc.configure_args.append("--disable-qt5")
+        tc.configure_args.append("--with-systemdsystemunitdir=/lib/systemd/system")
+        tc.generate()
```
Comment (uilianries) [inline recipes/avahi/all/conanfile.py:72]:
```suggestion
        tc.generate()
        tc = AutotoolsDeps(self)
        tc.generate()
        tc = PkgConfigDeps(self)
        tc.generate()
```

File: recipes/avahi/all/conanfile.py [hunk 9]
```diff
@@ -26,75 +30,68 @@ class AvahiConan(ConanFile):
...
+        copy(self, "LICENSE", self.source_folder, os.path.join(self.package_folder, "licenses"))
+        rmdir(self, os.path.join(self.package_folder, "etc"))
+        rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
+        rmdir(self, os.path.join(self.package_folder, "run"))
+        rmdir(self, os.path.join(self.package_folder, "share"))
+        rm(self, "*.la", os.path.join(self.package_folder, "lib"))
 
     def package_info(self):
         self.cpp_info.names["cmake_find_package"] = "Avahi"
         self.cpp_info.names["cmake_find_package_multi"] = "Avahi"
```
Comment (uilianries) [inline recipes/avahi/all/conanfile.py:91]: I think it could be removed. There is reference for avahi on its project. Only Hunter uses Avahi, but still, it's not official.

File: recipes/avahi/all/conanfile.py (hunk 8 above)
Comment (jwillikers) [reply recipes/avahi/all/conanfile.py:71]: Done.

File: recipes/avahi/all/conanfile.py (hunk 6 above)
Comment (jwillikers) [reply recipes/avahi/all/conanfile.py]: Done.

File: recipes/avahi/all/conanfile.py (hunk 5 above)
Comment (jwillikers) [reply recipes/avahi/all/conanfile.py]: Done.

File: recipes/avahi/all/conanfile.py (hunk 9 above)
Comment (jwillikers) [reply recipes/avahi/all/conanfile.py]: Done.

File: recipes/avahi/all/conanfile.py (hunk 2 above)
Comment (prince-chrismc) [reply recipes/avahi/all/conanfile.py:70]: Thanks for the amazing answer!

File: recipes/avahi/all/conanfile.py (hunk 3 above)
Comment (prince-chrismc) [reply recipes/avahi/all/conanfile.py:93]: Perfect

**Review by prince-chrismc — APPROVED** [review]

**Review by toge — APPROVED** [review]

File: recipes/avahi/all/conanfile.py [hunk 10]
```diff
@@ -26,75 +29,67 @@ class AvahiConan(ConanFile):
...
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version],
-                  destination=self._source_subfolder, strip_root=True)
-
-    @property
-    def _configure_args(self):
-        yes_no = lambda v: "yes" if v else "no"
-        return [
-            "--enable-shared={}".format(yes_no(self.options.shared)),
-            "--enable-static={}".format(yes_no(not self.options.shared)),
```
Comment (garethsb) [inline recipes/avahi/all/conanfile.py:62]: @jwillikers Why are these no longer necessary?

File: recipes/avahi/all/conanfile.py [hunk 11]
```diff
@@ -26,75 +29,67 @@ class AvahiConan(ConanFile):
...
+        rmdir(self, os.path.join(self.package_folder, "etc"))
+        rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
+        rmdir(self, os.path.join(self.package_folder, "run"))
+        rmdir(self, os.path.join(self.package_folder, "share"))
+        rm(self, "*.la", os.path.join(self.package_folder, "lib"))
 
     def package_info(self):
-        self.cpp_info.names["cmake_find_package"] = "Avahi"
-        self.cpp_info.names["cmake_find_package_multi"] = "Avahi"
-
```
Comment (garethsb) [inline recipes/avahi/all/conanfile.py:95]: @jwillikers Why were these removed?
The target names used in the original recipe were based on some common practice, e.g.
* https://github.com/sony/nmos-cpp/blob/master/Development/third_party/cmake/FindAvahi.cmake
* https://github.com/ruslo/hunter/blob/master/cmake/find/FindAvahi.cmake

File: recipes/avahi/all/conanfile.py (hunk 10 above)
Comment (garethsb) [reply recipes/avahi/all/conanfile.py:62]: Ah, handled automatically by `AutotoolsToolchain` - https://docs.conan.io/en/latest/reference/conanfile/tools/gnu/autotoolstoolchain.html#:~:text=enabled-shared

File: recipes/avahi/all/conanfile.py (hunk 11 above)
Comment (jwillikers) [reply recipes/avahi/all/conanfile.py:95]: Conan Center's policy, documented [here](https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/reviewing.md#cmake-targets), is to match target names of either those provided by CMake's own Find Modules or those officially provided by the library itself. When adding support for the newer Conan CMakeDeps generator, this was brought up. Instead of making the names inconsistent by following the policy for the CMakeDeps generator, it was determined to remove the previous aliases. You might try opening an issue in Conan Center to get feedback from the other Conan Center contributors on this.

Comment (garethsb) [reply recipes/avahi/all/conanfile.py:95]: Thanks, I really appreciate the quick reply. I opened #14223 but I have little hope of it being merged, despite that the uppercase named target `Avahi::Avahi` is common practice since at least 2015, because no-one who adopted it would be considered "official".
